### PR TITLE
Add data provider flags for FastExtent2D/FastExtent3D, simplify extent calculation for some providers

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR adds data provider flags for FastExtent2D/FastExtent3D. These flags reflect that retrieval of the provider's 2D or 3D extent retrieval via QgsDataProvider::extent()/extent3D() are **ALWAYS** guaranteed to be trivial/fast to calculate and involve absolutely no extra work.

Then, if the provider reflects these flags, we ALWAYS just retrieve the provider's extent directly when asked for a vector layer's 2d/3d extent. This not only considerably simplifies things for these providers (we can ignore lazy loaded extents and stored xml extents), but also ensures that the extent of these layers ALWAYS stays in sync with the actual data provider's known extents.

I've added the flags for all providers which meet this criteria (i.e. those where calls to extent() are just returning an already calculated member variable).

While I'm reluctant to add additional logic to the vector layer extent calculations, I think this is a valuable improvement for providers which meet this criteria. In particular, it can be used to solve an issue where a data provider has internal knowledge regarding changes to the data provider's actual extent, and there's no easy way for the provider to inform the layer of these changes (while there's been a fullExtentCalculated() signal in QgsDataProvider for many years, using signals for this is a risky approach that may cause deadlocks or other issues when eg a data provider knows that the extent has changed but is not ready for the map layer level slots connecting to this signal to be called immediately)

### CODE REVIEW

File: src/core/vector/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -1032,22 +1032,34 @@ QgsRectangle QgsVectorLayer::extent() const
   if ( !isSpatial() )
     return rect;
 
-  if ( !mValidExtent2D && mLazyExtent2D && mReadExtentFromXml && !mXmlExtent2D.isNull() )
+  if ( mDataProvider && mDataProvider->isValid() && mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D )
```
Comment (rouault) [inline src/core/vector/qgsvectorlayer.cpp:1035]: I'm always nervous about operator priority ...
```suggestion
  if ( mDataProvider && mDataProvider->isValid() && (mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D) )
```

File: src/core/vector/qgsvectorlayer.cpp [hunk 2]
```diff
@@ -1133,22 +1145,34 @@ QgsBox3D QgsVectorLayer:: extent3D() const
   if ( !isSpatial() )
     return extent;
 
-  if ( !mValidExtent3D && mLazyExtent3D && mReadExtentFromXml && !mXmlExtent3D.isNull() )
+  if ( mDataProvider && mDataProvider->isValid() && mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent3D )
```
Comment (rouault) [inline src/core/vector/qgsvectorlayer.cpp:1148]:
```suggestion
  if ( mDataProvider && mDataProvider->isValid() && (mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent3D) )
```

File: src/core/vector/qgsvectorlayer.cpp [hunk 3]
```diff
@@ -1032,22 +1032,34 @@ QgsRectangle QgsVectorLayer::extent() const
   if ( !isSpatial() )
     return rect;
 
-  if ( !mValidExtent2D && mLazyExtent2D && mReadExtentFromXml && !mXmlExtent2D.isNull() )
+  if ( mDataProvider && mDataProvider->isValid() && (mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D) )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayer.cpp:1035]:
```suggestion
  if ( mDataProvider && mDataProvider->isValid() && ( mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D ) )
```

File: src/core/vector/qgsvectorlayer.cpp [hunk 4]
```diff
@@ -1133,22 +1145,34 @@ QgsBox3D QgsVectorLayer:: extent3D() const
   if ( !isSpatial() )
     return extent;
 
-  if ( !mValidExtent3D && mLazyExtent3D && mReadExtentFromXml && !mXmlExtent3D.isNull() )
+  if ( mDataProvider && mDataProvider->isValid() && (mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent3D) )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayer.cpp:1148]:
```suggestion
  if ( mDataProvider && mDataProvider->isValid() && ( mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent3D ) )
```

**Review by rouault — APPROVED** [review]

**Pull request merged**